File: 27.移除元素.py
#
# @lc app=leetcode.cn id=27 lang=python
#
# [27] 移除元素
#
# 解法1(T95% S45%)：想法不难，难在边界的判断上。设定左右指针，左指针发现等于val的时候，右指针不断向前找第一个不等于val的值，此时如果左右指针同位了，直接返回当前位置，否则左右指针元素对调继续遍历，最终返回左指针的位置

# @lc code=start
class Solution(object):
    def removeElement(self, nums, val):
        """
        :type nums: List[int]
        :type val: int
        :rtype: int
        """
        left, right = 0, len(nums)-1
        while left <= right:
            if nums[left] == val:
                while right>left and nums[right]==val:
                    right -= 1
                if right==left: return left
                nums[left], nums[right] = nums[right], nums[left]
            left += 1

        return left
# @lc code=end

File: javascript/27.移除元素.js
/*
 * @lc app=leetcode.cn id=27 lang=javascript
 *
 * [27] 移除元素
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @param {number} val
 * @return {number}
 */
var removeElement = function(nums, val) {
    if (nums.length == 0) { return 0; }
    if (nums.length == 1) {
        if (nums[0] == val) { return 0; }
        else { return 1; }
    }

    var backNum = 0;
    var i=0;
    while (i<=nums.length - backNum - 1) {  // 注意这个=
        if (nums[i] == val) {
            var tmp = nums[i];
            nums[i] = nums[nums.length - backNum - 1];
            nums[nums.length - backNum - 1] = tmp;
            backNum++;
        } else {
            i++;
        }
    }
    return i;
};
// @lc code=end

